### PR TITLE
grpc_slice_refcount devirtualization

### PR DESCRIPTION
**Pull request by arjunroy**
De-virtualize grpc_slice_refcount.

---
This change is [Reviewable](https://reviewable.io/reviews/grpc/grpc/18407)

### CODE REVIEW

File: src/core/lib/slice/slice_intern.cc [hunk 1]
```diff
@@ -39,14 +41,7 @@
...
 
-typedef struct interned_slice_refcount {
-  grpc_slice_refcount base;
-  grpc_slice_refcount sub;
-  size_t length;
-  gpr_atm refcnt;
-  uint32_t hash;
-  struct interned_slice_refcount* bucket_next;
-} interned_slice_refcount;
+typedef grpc_core::InternedSliceRefcount interned_slice_refcount;
```
Comment (vjpai) [inline src/core/lib/slice/slice_intern.cc]: I can understand that this is to minimize the diff, but it looks odd to typedef something with C++-style naming/behavior to something with C-style naming/behavior.

File: src/core/lib/slice/slice_intern.cc [hunk 2]
```diff
@@ -56,7 +51,7 @@ typedef struct slice_shard {
 } slice_shard;
 
 /* hash seed: decided at initialization time */
-static uint32_t g_hash_seed;
+uint32_t g_hash_seed;
```
Comment (vjpai) [inline src/core/lib/slice/slice_intern.cc]: Great catch.

File: src/core/lib/slice/slice_intern.cc [hunk 3]
```diff
@@ -69,62 +64,27 @@ typedef struct {
...
+namespace grpc_core {
 
-static void interned_slice_destroy(interned_slice_refcount* s) {
-  slice_shard* shard = &g_shards[SHARD_IDX(s->hash)];
-  gpr_mu_lock(&shard->mu);
-  GPR_ASSERT(0 == gpr_atm_no_barrier_load(&s->refcnt));
+void InternedSliceRefcount::DestroyInstance() {
+  slice_shard* shard = &g_shards[SHARD_IDX(this->hash)];
+  MutexLock lock(&shard->mu);
+  // TODO(arjunroy) Really need this method
```
Comment (vjpai) [inline src/core/lib/slice/slice_intern.cc]: ??? Is this needed or deletable?

File: src/core/lib/slice/slice_intern.cc [hunk 4]
```diff
@@ -69,62 +64,27 @@ typedef struct {
...
+  // GPR_ASSERT(0 == gpr_atm_no_barrier_load(&this->refcnt));
   interned_slice_refcount** prev_next;
   interned_slice_refcount* cur;
-  for (prev_next = &shard->strs[TABLE_IDX(s->hash, shard->capacity)],
+  for (prev_next = &shard->strs[TABLE_IDX(this->hash, shard->capacity)],
```
Comment (vjpai) [inline src/core/lib/slice/slice_intern.cc]: Any reason for `this->` ? This isn't in a template declaration for example

File: src/core/lib/slice/slice_internal.h [hunk 5]
```diff
@@ -23,17 +23,168 @@
...
+    }
+  }
+
+  bool Validate() const { return ref_ != nullptr; }
+
+ private:
+  friend struct ::grpc_slice_refcount;
+
+  grpc_slice_refcount* sub_refcount_ = nullptr;
+  Type ref_type_ = REGULAR;
```
Comment (soheilhy) [inline src/core/lib/slice/slice_internal.h]: this should be const, and probably don't need an initializer.

File: src/core/lib/slice/slice_internal.h [hunk 6]
```diff
@@ -23,17 +23,168 @@
...
+  RefCount* ref_ = nullptr;
```
Comment (soheilhy) [inline src/core/lib/slice/slice_internal.h]: `ref_` seems to be accessed more frequently. If that's the case, I'd suggest pulling it up so that's in front of the cache-line.

File: src/core/lib/slice/slice_internal.h [hunk 7]
```diff
@@ -23,17 +23,168 @@
...
+}  // namespace grpc_core
+
+struct grpc_slice_refcount {
+  grpc_slice_refcount() { impl_.sub_refcount_ = this; }
+  grpc_slice_refcount(grpc_core::SliceRefcount::Type type,
+                      grpc_core::RefCount* ref, bool ensureRefNonZero,
+                      grpc_core::SliceRefcount::DestroyerFn destroyer_fn,
+                      void* destroyer_arg, grpc_slice_refcount* sub)
+      : impl_(type, ref, ensureRefNonZero, destroyer_fn, destroyer_arg, sub) {
+    // Nothing to do here.
```
Comment (soheilhy) [inline src/core/lib/slice/slice_internal.h]: nit: No need to add these comments.

File: src/core/lib/slice/slice_internal.h [hunk 8]
```diff
@@ -23,17 +23,168 @@
...
+  }
+  grpc_slice_refcount(grpc_slice_refcount* sub,
+                      grpc_core::SliceRefcount::Type type)
+      : impl_(sub, type) {
+    // Nothing to do here.
```
Comment (soheilhy) [inline src/core/lib/slice/slice_internal.h]: same here.

File: src/core/lib/slice/slice_internal.h [hunk 9]
```diff
@@ -23,17 +23,168 @@
...
+  InternedSliceRefcount(size_t length, uint32_t hash,
+                        InternedSliceRefcount* bucket_next)
+      : base(grpc_core::SliceRefcount::Type::INTERNED, &refcnt,
+             /*ensureRefNonZero=*/false, Destroy, this, &sub),
+        sub(grpc_core::SliceRefcount::Type::REGULAR, &refcnt,
+            /*ensureRefNonZero=*/false, Destroy, this, &sub),
+        length(length),
+        hash(hash),
+        bucket_next(bucket_next) {
+    // Nothing else to do here.
```
Comment (soheilhy) [inline src/core/lib/slice/slice_internal.h]: nit: please remove similar comments here and elsewhere.

Comment (soheilhy) [inline src/core/lib/slice/slice_internal.h]: nit: please remove similar comments here and elsewhere.

File: src/core/lib/slice/slice_internal.h [hunk 10]
```diff
@@ -23,17 +23,168 @@
...
+  }
+
+  void DestroyInstance();
+
+  grpc_slice_refcount base;
+  grpc_slice_refcount sub;
+  size_t length;
```
Comment (soheilhy) [inline src/core/lib/slice/slice_internal.h]: should `length` and `hash` be both const?

File: src/core/lib/slice/slice_internal.h [hunk 11]
```diff
@@ -23,17 +23,168 @@
...
+  switch (ref_type_) {
+    case STATIC:
+      return GRPC_STATIC_METADATA_INDEX(a) == GRPC_STATIC_METADATA_INDEX(b);
+    case INTERNED:
+      return a.refcount == b.refcount;
+    case REGULAR:
+      break;
+  }
+  if (GRPC_SLICE_LENGTH(a) != GRPC_SLICE_LENGTH(b)) return false;
+  if (GRPC_SLICE_LENGTH(a) == 0) return true;
```
Comment (soheilhy) [inline src/core/lib/slice/slice_internal.h]: for my education: why is this correct? can you add a comment please?

**Comment by arjunroy** [comment]
> That said... it was a little unexpected that changing from virtual functions to a switch would have had such a positive impact on performance; can you document or detail how that arises so that we won't have the temptation to switch it back the other direction? Are the branches in the switch highly predictable, for example? Is that because they're dominated by one specific type of use?

At a very high level, one example is https://godbolt.org/z/NJRt01

Example: Eq() with static metadata refcount.

New way: 16 instructions, 1 branch.
Old way: 26 instructions, 3 branches.

new_way:
main:
        sub     rsp, 56
        mov     eax, DWORD PTR ds:0
        mov     rdx, QWORD PTR [rsp]
        test    eax, eax
        je      .L2
.L2:
        xor     eax, eax
        sub     rdx, OFFSET FLAT:grpc_static_metadata_refcounts
        sub     rax, OFFSET FLAT:grpc_static_metadata_refcounts
        sar     rdx, 2
        sar     rax, 2
        cmp     eax, edx
        sete    al
        movzx   eax, al
        jmp     .L1
.L1:
        add     rsp, 56
        ret
Total: 16 instructions, 1 branch.

old way:
        sub     rsp, 104
        lea     rsi, [rsp+48]
        mov     rdi, rsp
        call    grpc_slice_eq(grpc_slice&, grpc_slice&)
grpc_slice_eq(grpc_slice&, grpc_slice&):
        mov     rax, QWORD PTR [rdi+8]
        test    rax, rax
        je      .L24
        mov     rdx, QWORD PTR [rsi+8]
        test    rdx, rdx
        je      .L24
        mov     rax, QWORD PTR [rax]
        cmp     rax, QWORD PTR [rdx]
        je      .L50
.L50:
        jmp     [QWORD PTR [rax+16]]
grpc_static_slice_eq(grpc_slice&, grpc_slice&):
        mov     rdx, QWORD PTR [rdi]
        mov     rax, QWORD PTR [rsi]
        sub     rdx, OFFSET FLAT:grpc_static_metadata_refcounts
        sub     rax, OFFSET FLAT:grpc_static_metadata_refcounts
        sar     rdx, 2
        sar     rax, 2
        cmp     edx, eax
        sete    al
        movzx   eax, al
        ret
        add     rsp, 104
        ret

Total: 26 instructions, 3 branches.

File: src/core/lib/slice/slice_internal.h [hunk 12]
```diff
@@ -23,17 +23,168 @@
...
+extern uint32_t g_hash_seed;
+namespace grpc_core {
+
+class SliceRefcount {
+ public:
+  enum Type { STATIC, INTERNED, REGULAR };
+  typedef void (*DestroyerFn)(void*);
+
+  SliceRefcount() = default;
+  SliceRefcount(SliceRefcount::Type type, RefCount* ref, bool ensureRefNonZero,
```
[the remark this answers is not in the document]
Comment (arjunroy) [reply src/core/lib/slice/slice_internal.h]: Fixed.

File: src/core/lib/slice/slice_internal.h [hunk 13]
```diff
@@ -23,17 +23,168 @@
...
+#include "src/core/lib/gprpp/ref_counted.h"
+#include "src/core/lib/transport/static_metadata.h"
+
+extern uint32_t static_metadata_hash_values[GRPC_STATIC_MDSTR_COUNT];
+extern uint32_t g_hash_seed;
+namespace grpc_core {
+
+class SliceRefcount {
+ public:
+  enum Type { STATIC, INTERNED, REGULAR };
```
[the remark this answers is not in the document]
Comment (arjunroy) [reply src/core/lib/slice/slice_internal.h]: Fixed.

File: src/core/lib/slice/slice_internal.h (hunk 8 above)
Comment (arjunroy) [reply src/core/lib/slice/slice_internal.h]: Removed

File: src/core/lib/slice/slice_internal.h (hunk 9 above)
Comment (arjunroy) [reply src/core/lib/slice/slice_internal.h]: Removed

Comment (arjunroy) [reply src/core/lib/slice/slice_internal.h]: Removed

File: src/core/lib/slice/slice_internal.h (hunk 10 above)
Comment (arjunroy) [reply src/core/lib/slice/slice_internal.h]: Agreed and changed.

File: src/core/lib/slice/slice_internal.h (hunk 11 above)
Comment (arjunroy) [reply src/core/lib/slice/slice_internal.h]: Actually there's two opportunities here.

Regarding the actual explanation, if two slices are of different lengths, they must be unequal (hence false). Else: same length. If that length is 0, trivially they are equal, so return true. Otherwise do the memcmp.

Opportunities:
#1: We can remove the len() == 0 check, since when 0 is passed into memcmp, we by default get 0 back.

#2: grpc_slice_eq by default is what people externally would use; that only uses this new Eq() method if the types are the same (in the old days: if the vtables were identical).
So if we compare, say, an interned slice to a regular slice, we'll call grpc_slice_default_eq_impl(grpc_slice a, grpc_slice b).

That is rather costly; it has the old style copy values invocation, and just does what we do in the fallthrough case of Eq() anyways.

So instead, I propose just using the new Eq() method always since it will always work even if the types are different.
ie. Suppose slice::a is static, but b is not. GRPC_STATIC_METADATA_INDEX(b) is just some pointer arithmetic followed by  a compare; the pointer itself is never deref'd. So that case is safe.
ie. Suppose slice::a is interned, but b is not. In general the only way two slices share the same refcount is if they're stream_refs or if they're interned. It should not be possible for a to be interned and for b to share the ref if b is not interned.

Then grpc_slice_default_eq_impl() still exists (being a part of the public API) but we can be faster by not using it. Thoughts?

Comment (soheilhy) [reply src/core/lib/slice/slice_internal.h]: Actually, never mind this comment. I forgot to resolve this. Thanks

File: src/core/lib/slice/slice_internal.h (hunk 5 above)
Comment (arjunroy) [reply src/core/lib/slice/slice_internal.h]: Made it const.

File: src/core/lib/slice/slice_internal.h [hunk 14]
```diff
@@ -23,17 +23,168 @@
...
+  Type getType() const { return ref_type_; }
+
+  int Eq(const grpc_slice& a, const grpc_slice& b);
+
+  uint32_t Hash(const grpc_slice& slice);
+  void Ref() {
+    if (!ref_) return;
+#ifndef NDEBUG
+    if (ensureRefNonZero_) {
+      ref_->RefNonZero();
```
[the remark this answers is not in the document]
Comment (arjunroy) [reply src/core/lib/slice/slice_internal.h]: Done.

File: src/core/lib/slice/slice_intern.cc (hunk 4 above)
Comment (arjunroy) [reply src/core/lib/slice/slice_intern.cc]: Readability; since that class hasn't been converted to _ suffixes on member vars this makes it a bit more obvious.

File: src/core/lib/slice/slice_intern.cc (hunk 3 above)
Comment (arjunroy) [reply src/core/lib/slice/slice_intern.cc]: @soheilhy thoughts on this?

Basically there's an assert here on destroy, which in the gpr_ref world we could manually check. But it's a private variable now.

Comment (arjunroy) [reply src/core/lib/slice/slice_intern.cc]: So for now I've commented out the check; to get it to work, we'd need to add a get() method to the grpc_core::RefCount.

File: src/core/lib/slice/slice_intern.cc (hunk 2 above)
Comment (arjunroy) [reply src/core/lib/slice/slice_intern.cc]: Yeah, needed the linkage change so I could inline the hash method for interned slices.

File: src/core/lib/slice/slice_intern.cc (hunk 1 above)
Comment (arjunroy) [reply src/core/lib/slice/slice_intern.cc]: I think for now it's an easier diff to review if we leave the names as is when able (so we don't have a million lines to change added)

File: src/core/lib/iomgr/resource_quota.cc [hunk 15]
```diff
@@ -430,41 +431,43 @@ static bool rq_reclaim(grpc_resource_quota* resource_quota, bool destructive) {
...
 static grpc_slice ru_slice_create(grpc_resource_user* resource_user,
                                   size_t size) {
-  ru_slice_refcount* rc = static_cast<ru_slice_refcount*>(
-      gpr_malloc(sizeof(ru_slice_refcount) + size));
-  rc->base.vtable = &ru_slice_vtable;
-  rc->base.sub_refcount = &rc->base;
-  gpr_ref_init(&rc->refs, 1);
-  rc->resource_user = resource_user;
-  rc->size = size;
+  using grpc_core::RuSliceRefcount;
```
[the remark this answers is not in the document]
Comment (arjunroy) [reply src/core/lib/iomgr/resource_quota.cc]: Removed.

File: src/core/lib/slice/slice.cc [hunk 16]
```diff
@@ -145,46 +140,49 @@ grpc_slice grpc_slice_new(void* p, size_t len, void (*destroy)(void*)) {
...
+};
 
-static const grpc_slice_refcount_vtable new_with_len_vtable = {
-    new_with_len_ref, new_with_len_unref, grpc_slice_default_eq_impl,
-    grpc_slice_default_hash_impl};
+}  // namespace grpc_core
 
 grpc_slice grpc_slice_new_with_len(void* p, size_t len,
                                    void (*destroy)(void*, size_t)) {
+  using grpc_core::New;
```
[the remark this answers is not in the document]
Comment (arjunroy) [reply src/core/lib/slice/slice.cc]: Removed.

File: src/core/lib/iomgr/resource_quota.cc [hunk 17]
```diff
@@ -430,41 +431,43 @@ static bool rq_reclaim(grpc_resource_quota* resource_quota, bool destructive) {
...
+ public:
+  static void Destroy(void* p) {
+    auto* rc = static_cast<RuSliceRefcount*>(p);
     grpc_resource_user_free(rc->resource_user, rc->size);
     gpr_free(rc);
   }
-}
+  RuSliceRefcount(grpc_resource_user* resource_user, size_t size)
+      : base(grpc_core::SliceRefcount::Type::REGULAR, &refs,
+             /*ensureRefNonZero=*/false, Destroy, this, &base),
```
[the remark this answers is not in the document]
Comment (arjunroy) [reply src/core/lib/iomgr/resource_quota.cc]: Per Soheil's comment, this is now removed.

File: src/core/lib/slice/slice_internal.h (hunk 6 above)
Comment (arjunroy) [reply src/core/lib/slice/slice_internal.h]: Done. Now Ref and Type are at the head.

**Comment by arjunroy** [comment]
Updated PR based on earlier comments. Still a couple of things outstanding (most notably, the get() => ASSERT check that we can no longer do).

File: src/core/lib/slice/slice_intern.cc (hunk 3 above)
Comment (soheilhy) [reply src/core/lib/slice/slice_intern.cc]: I would say this check is unnecessary, because it's an extra load, which is simply there for debugging. To elaborate, if we want to check that refcount is 0 when we are in destroy we would have to do that everywhere in gRPC. So, I would just drop it.

But I understand the need for get. We can add it to the ref count for debug builds.

Comment (arjunroy) [reply src/core/lib/slice/slice_intern.cc]: Ack; will just drop it.

File: src/core/ext/filters/client_channel/client_channel.cc [hunk 18]
```diff
@@ -252,11 +252,11 @@ class ClientChannelControlHelper
 // Synchronous callback from chand->resolving_lb_policy to process a resolver
 // result update.
 static bool process_resolver_result_locked(
-    void* arg, const grpc_channel_args& args, const char** lb_policy_name,
+    void* arg, grpc_core::Resolver::Result* result, const char** lb_policy_name,
```
Comment (vjpai) [inline src/core/ext/filters/client_channel/client_channel.cc]: ??? Unrelated change? Maybe the merge commit history is broken? May need some clever rebasing if possible

File: src/core/ext/filters/client_channel/client_channel.cc [hunk 19]
```diff
@@ -252,11 +252,11 @@ class ClientChannelControlHelper
...
     grpc_core::RefCountedPtr<LoadBalancingPolicy::Config>* lb_policy_config) {
   channel_data* chand = static_cast<channel_data*>(arg);
   chand->have_service_config = true;
-  ProcessedResolverResult resolver_result(args, chand->enable_retries);
+  ProcessedResolverResult resolver_result(result, chand->enable_retries);
```
Comment (vjpai) [inline src/core/ext/filters/client_channel/client_channel.cc]: Similar issue here

Comment (vjpai) [reply src/core/ext/filters/client_channel/client_channel.cc]: and looks like in many of the client_channel files

Comment (vjpai) [reply src/core/ext/filters/client_channel/client_channel.cc]: and tons of conflicts exist now

**Comment by arjunroy** [comment]
Accidental close.

File: src/core/ext/filters/client_channel/client_channel.cc (hunk 19 above)
Comment (arjunroy) [reply src/core/ext/filters/client_channel/client_channel.cc]: Fixed issue by rebasing.
